projects/canopy/src/lib/forms/sort-code/sort-code.component.ts

```diff
@@ -0,0 +1,206 @@
+      this.ariaDescribedBy,
+      this._validationElement,
+      element,
+    );
+    this._validationElement = element;
+  }
+
+  @ViewChild('sortCodeFormDirective')
+  formGroupDirective: FormGroupDirective;
+
+  constructor(
+    private domService: LgDomService,
+    private errorState: LgErrorStateMatcher,
+    @Self()
+    @Optional()
+    private ngControl: NgControl,
+    private parentFormGroupDirective: FormGroupDirective,
+  ) {
+    if (this.ngControl != null) {
+      this.ngControl.valueAccessor = this;
+    }
+    this.first = new FormControl(null, [
+      Validators.required,
+      Validators.pattern(/^\d\d$/),
+    ]);
+    this.second = new FormControl(null, [
+      Validators.required,
+      Validators.pattern(/^\d\d$/),
+    ]);
+    this.third = new FormControl(null, [
+      Validators.required,
+      Validators.pattern(/^\d\d$/),
+    ]);
+    this.sortCodeFormGroup = new FormGroup(
+      {
+        first: this.first,
+        second: this.second,
+        third: this.third,
+      },
+      {
+        updateOn: 'blur',
+      },
+    );
+  }
+
+  ngOnInit(): void {
+    // append internal validators to external validators if applicable.
+    if (this.ngControl && this.ngControl.control) {
+      this.ngControl.control.setValidators([this.validate.bind(this)]);
+      this.ngControl.control.updateValueAndValidity();
+    }
+
+    this.valueChanges = this.sortCodeFormGroup.valueChanges.subscribe(
+      (sortCode: SortCodeField) => {
+        const first = sortCode.first || '';
+        const second = sortCode.second || '';
+        const third = sortCode.third || '';
+        this.onChange(`${first}${second}${third}`);
+      },
+    );
+
+    // submit the group when the parent form is submitted
+    this.parentFormGroupDirective.ngSubmit
+      .pipe(filter(({ type }) => type === 'submit'))
+      .subscribe((event) => {
+        this.formGroupDirective.onSubmit(event);
+        this.ngControl.control.updateValueAndValidity();
+      });
+  }
+
+  isControlInvalid(control: NgControl, form: FormGroupDirective) {
+    return this.errorState.isControlInvalid(control, form);
+  }
+
+  writeValue(sortCode: string): void {
+    if (!sortCode) {
+      return;
+    }
+
+    this.sortCodeFormGroup.setValue({
+      first: sortCode.substr(0, 2),
+      second: sortCode.substr(2, 2),
+      third: sortCode.substr(4, 2),
+    });
+  }
+
+  validate(): ValidationErrors {
+    const errors: ValidationErrors = {};
+    Object.keys(this.sortCodeFormGroup.controls).forEach((fieldName) => {
+      if (
+        this.sortCodeFormGroup.controls[fieldName].invalid &&
+        (this.sortCodeFormGroup.controls[fieldName].dirty ||
+          (this.formGroupDirective && this.formGroupDirective.submitted))
+      ) {
+        errors.invalidField = true;
+      }
+    });
+    if (Object.keys(errors).length) {
+      return errors;
+    }
+    return null;
+  }
+
+  onBlur() {
```

Comment:
Could you add the return type please?

Reply:
Ah yes, will do 👍

Reply:
Done ✅

Reply:
would this enforce us to add return types?
https://stackoverflow.com/questions/42793701/is-there-a-way-to-enforce-method-return-types-on-typescript-classes-via-a-tslint

Reply:
Yes, there is definitely a rule. I think it's worth doing it when moving to eslint though as it might need quite a bit of refactoring.